Change HAL pin type for hy_gt_vfd.fault-info from float to u32

petterreinholdtsen opened the pull request:
Restructure code to handle two types of HAL pin types for modbus register.

Fixes #2206.

petterreinholdtsen commented:
I would like someone who know this component a little better to see if this is a wanted change and if the approach taken is a good one.

andypugh commented:
Looks reasonable, though possibly the input_terminal should be int too. (but then I suspect that output_terminal isn't really handled at all in any useful way)

SebKuzminsky commented:
Looks good to me.  I no longer have access to this VFD to test it, I'm interested to see if it works for @persei802.

I also don't have access to the manual, but I wonder if the fault-info register is a bitmap of different errors; if so it might make sense to break it out into a bunch of bit pins identifying the different errors.

persei802 commented:
That is correct. There are 26 bits indicating a fault condition. In my GUI, I convert this register to a hex value so it's easy to determine which bits are set. The manual doesn't actually explain what each of these bits mean, except that 0x00 is no fault. In my personal opinion, it is not useful to break them out into individual HAL pins.
Jim

SebKuzminsky commented on docs/src/man/man1/hy_gt_vfd.1.adoc:
```diff
@@ -257,9 +257,11 @@ PC.03 = 1::
+*hy_gt_vfd.fault-info* (u32, out)::
 
-        The VFD's fault info register.
+        The VFD's fault info register.  The type of this register was
+        changed from float to u32 in version 2.10.  The commit was
+        done 2022-12-17.
```
This kind of user-facing changelog information belongs in [Updating LinuxCNC](http://linuxcnc.org/docs/devel/html/getting-started/updating-linuxcnc.html), not in the manpage.

SebKuzminsky replied:
The patch was reworked to not break backward compatibility, so no note is needed in `Updating LinuxCNC`.

SebKuzminsky reviewed (changes requested):
Looks good, except for the nitpick about changelog location.

c-morley commented:
Could this be done with s32 rather then u32 or do you need to extra range? 
u32 is awkward to use in many components ie classicladder.

andypugh commented:
I think U32 is the right choice if the intention is to break it out into bits. It matches the bitslice component. 
It's a bitmap, so signed does not make any sense (even if it is only 16 bits)

c-morley commented:
Then maybe have both pins? HAL pins are cheap.

petterreinholdtsen commented:
[Sebastian Kuzminsky]
> @SebKuzminsky requested changes on this pull request.
>
> Looks good, except for the nitpick about changelog location.
>
>> +        The VFD's fault info register.  The type of this register was
> +        changed from float to u32 in version 2.10.  The commit was
> +        done 2022-12-17.
>
> This kind of user-facing changelog information belongs in [Updating
> LinuxCNC](http://linuxcnc.org/docs/devel/html/getting-started/updating-linuxcnc.html),
> not in the manpage.

The idea is to make sure those using the fault-info value today who
experience unexpected behaviour when they upgrade to 2.10, should be
able to find a clue why in the man page of the failing module.

I find it valuable when information about when a feature was introduced
is included in manual pages, like it is for example in several linux
syscall man pages and debhelper(7).

I have nothing against listing it in
getting-started/updating-linuxcnc.adoc.

-- 
Happy hacking
Petter Reinholdtsen

petterreinholdtsen commented:
[c-morley]
> Then maybe have both pins? HAL pins are cheap.

I left the old fault-info float pin alone, and introduced a new
fault-info-bits pin instead, to stay backwards compatible while also
providing a u32 with easier interpretation of the error bits.  Updated
documentation to match.

-- 
Happy hacking
Petter Reinholdtsen

c-morley commented:
Ahh I see, that works too.

persei802 commented:
I misspoke earlier. This register has 26 fault codes, not bits. It is a 16 bit register. Anyway, I ran Petter's modified code and it works fine. I have no further issues with this component.

petterreinholdtsen commented:
[Jim]
> I misspoke earlier. This register has 26 fault codes, not bits. It is
> a 16 bit register. Anyway, I ran Petter's modified code and it works
> fine. I have no further issues with this component.

Aha.  Adjusted documentation to drop the 'bit' claim.  Where is the list
of fault codes documnented?

-- 
Happy hacking
Petter Reinholdtsen

persei802 commented:
They are in the GT Series Inverter Manual.
![fault_codes](https://user-images.githubusercontent.com/56328463/208323864-373017fe-476d-4236-9ad0-8dc03cba6c16.png)

petterreinholdtsen commented:
[Jim]
> They are in the GT Series Inverter Manual.

Thank you.  Added this reference to the man page, and renamed the new
u32 pin from fault-info-bits to fault-info-code, as it is not a bit
field.

I believe this is ready for merging.
-- 
Happy hacking
Petter Reinholdtsen
